### PR TITLE
Allow affix settings to be dynamic / updatable

### PR DESCRIPTION
**Pull request by s1monw**
Today affix settings are not dynamic since it's required to know
it's namespace in order to pull a concrete setting from it. This is not possible
in practice since the namespaces are dynamic by design. This change allows to register
a specialized settings consumer that consumes the namespace and the actual value if
a setting gets updated.

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 1]
```diff
@@ -449,6 +452,76 @@ public String toString() {
         };
     }
 
+    public static class AffixSetting<T> extends Setting<T> {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/Setting.java]: Probably worth javadocs.

File: core/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 2]
```diff
@@ -1052,6 +1095,17 @@ String getConcreteString(String key) {
             return matcher.group(1);
         }
 
+        /**
+         * Returns a string representation of the concrete setting key
+         */
+        String getNamesapce(String key) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/Setting.java]: s/Namesapce/Namespace/

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 3]
```diff
@@ -126,6 +129,51 @@ public void testAddConsumer() {
         assertEquals(0, consumer2.get());
     }
 
+    public void testAddConsumerAffix() {
+        Setting.AffixSetting<Integer> testSetting = Setting.affixKeySetting("foo.", "bar",
+            (k) ->  Setting.intSetting(k, 1, Property.Dynamic, Property.NodeScope));
+        Setting.AffixSetting<List<Integer>> testSetting2 = Setting.affixKeySetting("foo.", "list",
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: name it `listSetting`?

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 4]
```diff
@@ -126,6 +129,51 @@ public void testAddConsumer() {
...
+            (k) -> Setting.listSetting(k, Arrays.asList("1"), Integer::parseInt, Property.Dynamic, Property.NodeScope));
+        AbstractScopedSettings service = new ClusterSettings(Settings.EMPTY,new HashSet<>(Arrays.asList(testSetting, testSetting2)));
+        Map<String, List<Integer>> listResults = new HashMap<>();
+        Map<String, Integer> intResults = new HashMap<>();
+
+        BiConsumer<String, List<Integer>> listConsumer = listResults::put;
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: I'd switch the order of these so it matches the declaration order.

File: core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java [hunk 5]
```diff
@@ -126,6 +129,51 @@ public void testAddConsumer() {
...
+        service.addAffixUpdateConsumer(testSetting, intConsumer, (s, k) -> {});
+        assertEquals(0, listResults.size());
+        assertEquals(0, intResults.size());
+        service.applySettings(Settings.builder()
+            .put("foo.test.bar", 2)
+            .put("foo.test_1.bar", 7)
+            .putArray("foo.test_list.list", "16", "17")
+            .putArray("foo.test_list_1.list", "18", "19", "20")
+            .build());
+        assertEquals(2, intResults.get("test").intValue());
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java]: Very cute.

**Review by nik9000 — APPROVED** [review]
LGTM. I left a few nit picks but nothing interesting. It'd be nice to have a non-test consumer for `addAffixUpdateConsumer` but isn't required for now.

**Comment by s1monw** [comment]
> LGTM. I left a few nit picks but nothing interesting. It'd be nice to have a non-test consumer for addAffixUpdateConsumer but isn't required for now.

cool thx, I will use this in #22502 but I didn't wanna make it part of that

**Pull request merged**